extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveTableCatalog.scala

```diff
@@ -0,0 +1,410 @@
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+import org.apache.kyuubi.spark.connector.hive.HiveTableCatalog.{toCatalogDatabase, CatalogDatabaseHelper, IdentifierHelper, NamespaceHelper}
+
+/**
+ * A [[TableCatalog]] that wrap HiveExternalCatalog to as V2 CatalogPlugin instance to access Hive.
+ */
+class HiveTableCatalog(sparkSession: SparkSession)
+  extends TableCatalog with SQLConfHelper with SupportsNamespaces with Logging {
+
+  def this() = this(SparkSession.active)
+
+  private val sc = sparkSession.sparkContext
+
+  private val sessionState = sparkSession.sessionState
+
+  private var catalogName: String = _
+
+  private var catalogOptions: CaseInsensitiveStringMap = _
+
+  private var catalog: HiveSessionCatalog = _
+
+  val NAMESPACE_RESERVED_PROPERTIES =
+    Seq(
+      SupportsNamespaces.PROP_COMMENT,
+      SupportsNamespaces.PROP_LOCATION,
+      SupportsNamespaces.PROP_OWNER)
+
+  private lazy val hadoopConf: Configuration = {
+    val conf = sparkSession.sessionState.newHadoopConf()
+    catalogOptions.asScala.foreach {
+      case (k, v) => conf.set(k, v)
+      case _ =>
+    }
+    conf
+  }
+
+  private lazy val sparkConf: SparkConf = {
+    val conf = sparkSession.sparkContext.getConf
+    catalogOptions.asScala.foreach {
+      case (k, v) => conf.set(k, v)
+    }
+    conf
+  }
+
+  def hadoopConfiguration(): Configuration = hadoopConf
+
+  override def name(): String = {
+    require(catalogName != null, "The Hive table catalog is not initialed")
+    catalogName
+  }
+
+  override def initialize(name: String, options: CaseInsensitiveStringMap): Unit = {
+    assert(catalogName == null, "The Hive table catalog is already initialed.")
+    assert(
+      conf.getConf(CATALOG_IMPLEMENTATION) == "hive",
+      s"Require setting ${CATALOG_IMPLEMENTATION.key} to `hive` to enable hive support.")
+    catalogName = name
+    catalogOptions = options
+    catalog = new HiveSessionCatalog(
+      externalCatalogBuilder = () => externalCatalog,
+      globalTempViewManagerBuilder = () => sparkSession.sharedState.globalTempViewManager,
+      metastoreCatalog = new HiveMetastoreCatalog(sparkSession),
+      functionRegistry = sessionState.functionRegistry,
+      tableFunctionRegistry = sessionState.tableFunctionRegistry,
+      hadoopConf = hadoopConf,
+      parser = sessionState.sqlParser,
+      functionResourceLoader = sessionState.resourceLoader)
+  }
+
+  /**
+   * A catalog that interacts with external systems.
+   */
+  lazy val externalCatalog: ExternalCatalogWithListener = {
+    val externalCatalog = new HiveExternalCatalog(sparkConf, hadoopConf)
+
+    // Wrap to provide catalog events
+    val wrapped = new ExternalCatalogWithListener(externalCatalog)
+
+    // Make sure we propagate external catalog events to the spark listener bus
+    wrapped.addListener((event: ExternalCatalogEvent) => postExternalCatalogEvent(sc, event))
```

Comment:
Good consideration, but should we need to define a new Event subclass to identify which catalog the event came from ?

Reply:
That's great, but seems `ExternalCatalogWithListener` post spark event(is subclass of ExternalCatalogEvent), not custom event, unless we rewrite `ExternalCatalogWithListener` to post our event, which seems a bit expensive (if it's just for signs), probably